include/verilated_saif_c.h

```diff
@@ -0,0 +1,430 @@
+    VL_ATTR_ALWINLINE void updateLastTime(uint64_t val) { m_lastTime = val; }
+
+    // ACCESSORS
+    VL_ATTR_ALWINLINE uint32_t getWidth() const { return m_width; }
+    VL_ATTR_ALWINLINE ActivityBit& getBit(std::size_t index);
+    VL_ATTR_ALWINLINE uint64_t getLastUpdateTime() const { return m_lastTime; }
+
+private:
+    // CONSTRUCTORS
+    VL_UNCOPYABLE(ActivityVar);
+
+    // MEMBERS
+    uint64_t m_lastTime{0};
+    ActivityBit* m_bits;
+    uint32_t m_width;
+};
+
+//=============================================================================
+// ActivityScope
+
+class ActivityScope final {
+public:
+    // CONSTRUCTORS
+    ActivityScope(std::string name, int32_t parentScopeIndex = -1)
+        : m_scopeName{std::move(name)}
+        , m_parentScopeIndex{parentScopeIndex} {}
+
+    ActivityScope(ActivityScope&&) = default;
+    ActivityScope& operator=(ActivityScope&&) = default;
+
+    // METHODS
+    VL_ATTR_ALWINLINE void addChildScopeIndex(int32_t index) {
+        m_childScopesIndices.emplace_back(index);
+    }
+    VL_ATTR_ALWINLINE void addActivityVar(uint32_t code, std::string name) {
+        m_childActivities.emplace_back(code, std::move(name));
+    }
+    VL_ATTR_ALWINLINE bool hasParent() const { return m_parentScopeIndex >= 0; }
+
+    // ACCESSORS
+    VL_ATTR_ALWINLINE const std::string& getName() const { return m_scopeName; }
+    VL_ATTR_ALWINLINE const std::vector<int32_t>& getChildScopesIndices() const {
+        return m_childScopesIndices;
+    }
+    VL_ATTR_ALWINLINE const std::vector<std::pair<uint32_t, std::string>>&
+    getChildActivities() const {
+        return m_childActivities;
+    }
+    VL_ATTR_ALWINLINE int32_t getParentScopeIndex() const { return m_parentScopeIndex; }
+
+private:
+    // CONSTRUCTORS
+    VL_UNCOPYABLE(ActivityScope);
+
+    // MEMBERS
+    std::string m_scopeName{};
+    std::vector<int32_t> m_childScopesIndices{};
+    std::vector<std::pair<uint32_t, std::string>> m_childActivities{};
+    int32_t m_parentScopeIndex{-1};
+};
+
+//=============================================================================
+// VerilatedSaif
+// Base class to create a Verilator SAIF dump
+// This is an internally used class - see VerilatedSaifC for what to call from applications
+
+class VerilatedSaif VL_NOT_FINAL : public VerilatedTrace<VerilatedSaif, VerilatedSaifBuffer> {
+public:
+    using Super = VerilatedTrace<VerilatedSaif, VerilatedSaifBuffer>;
+
+private:
+    friend VerilatedSaifBuffer;  // Give the buffer access to the private bits
+
+    //=========================================================================
+    // SAIF-specific internals
+
+    VerilatedSaifFile* m_filep;  // File we're writing to
+    bool m_fileNewed;  // m_filep needs destruction
+    bool m_isOpen = false;  // True indicates open file
+    std::string m_filename;  // Filename we're writing to (if open)
+    uint64_t m_rolloverSize = 0;  // File size to rollover at
```

Comment:
As far as I understand the size of a SAIF file is constant, proportional to the number of bits it contains, so I'm not sure what rollover is used for here. Can we remove it? See the FST implementation for reference which does not have rollover.

Reply:
I agree size-based rollover unlikely makes sense.  Sometimes power methodologies do want to e.g. sample every 1us or something, if so it might make sense to support time-triggered rollover instead.